k8s/secrets.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2019 OpenFaaS Authors
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+
+	types "github.com/openfaas/faas-provider/types"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	typedV1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const (
+	secretsMountPath = "/var/openfaas/secrets"
+	secretLabel      = "app.kubernetes.io/managed-by"
+	secretLabelValue = "openfaas"
+)
+
+// SecretsClient exposes the standardized CRUD behaviors for Kubernetes secrets.  These methods
+// will ensure that the secrets are structured and labelled correctly for use by the OpenFaaS system.
+type SecretsClient interface {
+	// List returns a list of available function secrets.  Only the names are returned
+	// to ensure we do not accidentally read or print the sensitive values during
+	// read operations.
+	List(ctx context.Context, namespace string) (names []string, err error)
+	// Create adds a new secret, with the appropriate labels and structure to be
+	// used as a function secret.
+	Create(ctx context.Context, secret types.Secret) error
+	// Replace updates the value of a function secret
+	Replace(ctx context.Context, secret types.Secret) error
+	// Delete removes a function secret
+	Delete(ctx context.Context, name string, namespace string) error
+	// GetSecrets queries Kubernetes for a list of secrets by name in the given k8s namespace.
+	// This should only be used if you need access to the actual secret structure/value. Specifically,
+	// inside the FunctionFactory.
+	GetSecrets(ctx context.Context, namespace string, secretNames []string) (map[string]*apiv1.Secret, error)
+}
+
+// SecretsInterfacer exposes the SecretInterface getter for the k8s client.
+// This is implemented by the CoreV1Interface() interface in the Kubernetes client.
+// The SecretsClient only needs this one interface, but needs to be able to set the
+// namespaces when the interface is instantiated, meaning, we need the Getter and not the
+// SecretInterface itself.
+type SecretInterfacer interface {
+	// Secrets returns a SecretInterface scoped to the specified namespace
+	Secrets(namespace string) typedV1.SecretInterface
+}
+
+type secretClient struct {
+	kube SecretInterfacer
+}
+
+// NewSecretsClient constructs a new SecretsClient using the provided Kubernetes client.
+func NewSecretsClient(kube kubernetes.Interface) SecretsClient {
+	return &secretClient{
+		kube: kube.CoreV1(),
+	}
+}
+
+func (c secretClient) List(ctx context.Context, namespace string) (names []string, err error) {
```

Comment:
Where is `ctx` being used? I can't seem to find it.

Reply:
It isn't used, it is a work habit since this would normally be a method that should be traced (because it has IO) but we don't have any tracing support at the moment, so i will remove it